Dependency caching blog WIP (#13302)
pulumi/docs: content/blog/announcing-dependency-caching-deployments/index.md

meta_image: meta.png
authors:
- komal-ali
- iaroslav-titov
tags:
- features
- pulumi-deployments
- caching
---

We’re excited to announce dependency caching for [Pulumi Deployments](/docs/pulumi-cloud/deployments/)! This new feature is designed to significantly speed up deployment times by reducing dependency installation time by up to 80%, ensuring faster and more efficient workflows for every Pulumi user.

When using Pulumi Deployments, installing dependencies can be a time-consuming part of a deployment. With dependency caching, Pulumi can reuse previously installed dependencies, allowing your deployments to skip redundant installation steps and focus on the task at hand. This optimization is especially helpful for frequent deployments, reducing both time and compute resources.

<!--more-->

## How dependency caching works

Pulumi Deployments now automatically caches dependencies for popular package managers, such as npm, pip, and go modules. When a deployment is triggered, Pulumi will:

1.	Check for an existing cache of your dependencies.
2.	If a cache already exists, Pulumi will reuse those dependencies, bypassing the need to install them from scratch.

> 3.	If a cache does not exist, or if dependencies have changed, Pulumi will manage the cache to be used for subsequent deployments.

Reviewer: ```suggestion
3.	If a cache does not exist, or if dependencies have changed, Pulumi will install fresh dependences and create an updated cache to be used for subsequent deployments.
```

The end result? Faster, more efficient deployments to push out infrastructure changes quickly.

## How to enable dependency caching

Dependency caching can be enabled through deployment settings, however you choose to configure them.

### In the Pulumi Console

1. Open the Pulumi Console and navigate to a stack's deployment settings (i.e. Stack page > Settings > Deploy).
2. Under the "Dependency caching" section, toggle the switch to enable dependency caching for your stack.
3. Click "Save deployment configuration" to apply the changes.

![Dependency caching in the Pulumi Console](enable-caching-console.png)

### Using the REST API

> You can also enable dependency caching using the Pulumi REST API. See the [API documentation](/docs/pulumi-cloud/deployments/api/#cache-options) for more information.

Reviewer: `"You can also ..."` reads to me as the final item in a list of options, but then its followed up by PSP.  Maybe move the REST API item to the end?

### Defined as code with the Pulumi Cloud Service provider

If you prefer to define your deployment settings as code, you can use the Pulumi Cloud Service provider to enable dependency caching. Here’s an example of how to do this in TypeScript:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as service from "@pulumi/pulumiservice";

const config = new pulumi.Config();

const settings = new service.DeploymentSettings("deployment_settings", {
    organization: "service-provider-test-org",
    project: "test-deployment-settings-project",
    stack: "dev",
    operationContext: {
        environmentVariables: {
            TEST_VAR: "foo",
            SECRET_VAR: config.requireSecret("my_secret"),
        }
    },
    sourceContext: {
        git: {
            repoUrl: "https://github.com/pulumi/deploy-demos.git",
            branch: "refs/heads/main",
            repoDir: "pulumi-programs/simple-resource"
        }
    },
    cacheOptions: {

>         enable: true,

Reviewer: I think putting a comment on the line that actually enables the feature would be nice, eg
```suggestion
        enable: true, // <--
```

---
title: "Announcing dependency caching for Pulumi Deployments"

> date: 2024-11-11

Reviewer 2: Updated timestamp for this to show above the CosmosDB post

meta_desc: "Dependency caching for Pulumi Deployments is now available, speeding up your deployments by reusing dependencies across runs."
meta_image: meta.png
authors:
- komal-ali
- iaroslav-titov
tags:
- features
- pulumi-deployments
- caching
---

> We’re excited to announce dependency caching for Pulumi Deployments! This new feature is designed to significantly speed up deployment times by caching dependencies between runs, ensuring faster and more efficient workflows for every Pulumi user.

Reviewer 2: In the best case, can it be 75% faster? If yes, would be good to call out.

When using Pulumi Deployments, installing dependencies can be a time-consuming part of a deployment. With dependency caching, Pulumi can reuse previously installed dependencies, allowing your deployments to skip redundant installation steps and focus on the task at hand. This optimization is especially helpful for frequent deployments, reducing both time and compute resources.

<!--more-->

## How dependency caching works

Pulumi Deployments now automatically caches dependencies for popular package managers, such as npm, pip, and go modules. When a deployment is triggered, Pulumi will:

1.	Check for an existing cache of your dependencies.

> 2.	If a cache is available and relevant to your project, Pulumi will reuse those dependencies, bypassing the need to install them from scratch.

Reviewer 3: `available and relevant to your project` sounds a bit too wordy, maybe `If a cache already exists` or similar?

Also maybe change `If dependencies have changed` to `Otherwise, or if dependencies have changed` + to make sure people understand how cache gets created/not just updated, but this one might be hard to reword the whole sentence

> 3.	If a cache does not exist, or if dependencies have changed, Pulumi will create a new cache to be used for subsequent deployments.

Reviewer 4: ```suggestion
3.	If a cache does not exist, or if dependencies have changed, Pulumi will manage the cache to be used for subsequent deployments.
```

> The end result? Faster, more efficient deployments that get your infrastructure changes live more quickly.

Reviewer 4: Maybe it is just me but I read the other form of "live" than intended and had to read it twice.

Author: "deployed more quickly"?

Reviewer 4: Maybe "The end result? Faster, more efficient deployments to push out infrastructure changes quickly."

## How to enable dependency caching

> Dependency caching can be enabled through deployment settings, however you choose to configure them.

Reviewer 2: ```suggestion
Dependency caching can be enabled in numerous ways: 
```

Reviewer 3: I kinda like the current version better

### Using the REST API

You can also enable dependency caching using the Pulumi REST API. See the [API documentation](docs/pulumi-cloud/deployments/api/#cache-options) for more information.

### Defined as code with the Pulumi Cloud Service provider

If you prefer to define your deployment settings as code, you can use the Pulumi Cloud Service provider to enable dependency caching. Here’s an example of how to do this in TypeScript:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as service from "@pulumi/pulumiservice";

const config = new pulumi.Config();

const settings = new service.DeploymentSettings("deployment_settings", {
    organization: "service-provider-test-org",
    project: "test-deployment-settings-project",
    stack: "dev",
    operationContext: {
        environmentVariables: {
            TEST_VAR: "foo",
            SECRET_VAR: config.requireSecret("my_secret"),
        }
    },
    sourceContext: {
        git: {
            repoUrl: "https://github.com/pulumi/deploy-demos.git",
            branch: "refs/heads/main",
            repoDir: "pulumi-programs/simple-resource"
        }
    },
    cacheOptions: {
        enable: true,
    }
});
```

> If your project uses a supported package manager, Pulumi will automatically detect and manage dependency caching for you.

Reviewer 4: Where do we list "supported package managers"?

Author: In the FAQ but this could maybe be clearer.

Reviewer 4: Yeah, with the link to the FAQ in the next paragraph, I may be left wondering where the approved list is.

> For more information on how dependency caching works with Pulumi Cloud Deployments, check out the [FAQ](/docs/pulumi-cloud/deployments/faq/#dependency-caching) on dependency caching.

Reviewer 4: Do we have a PR to update that FAQ page?

Author: Is there information that you think should be there that isn't currently? https://www.pulumi.com/docs/pulumi-cloud/deployments/faq/#dependency-caching

Reviewer 4: Didn't know that was live already, disregard!
